### PR TITLE
add categorical values for heatmap (fix for JuliaPlots/Makie.jl#400

### PR DESCRIPTION
**Pull request by SimonDanisch**
Fixes JuliaPlots/Makie.jl#400

### CODE REVIEW

**Comment by asinghvi17** [comment]
Shall we also add a test to MakieGallery for this?

**Comment by SimonDanisch** [comment]
Would be great!

**Comment by asinghvi17** [comment]
Strange - it looks like I'm getting an error in the OpenGL conversion:

```julia
julia> x = ["a", "b", "c"]
3-element Array{String,1}:
 "a"
 "b"
 "c"

julia> y = ["α", "β", "γ"]
3-element Array{String,1}:
 "α"
 "β"
 "γ"

julia> heatmap(x, y, rand(3, 3))
Error showing value of type Scene:
ERROR: MethodError: no method matching gl_convert(::Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}})
Closest candidates are:
  gl_convert(::Nothing) at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLUniforms.jl:202
  gl_convert(::GLMakie.GLAbstraction.GLProgram, ::Any) at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLShader.jl:114
  gl_convert(::GLMakie.GLAbstraction.AbstractLazyShader, ::Any) at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLShader.jl:221
  ...
Stacktrace:
 [1] lift(::Function, ::Observables.Observable{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}) at /Users/anshul/.julia/dev/AbstractPlotting/src/interaction/nodes.jl:9
 [2] const_lift at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLUtils.jl:197 [inlined]
 [3] gl_convert(::Observables.Observable{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}) at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLUniforms.jl:230
 [4] GLMakie.GLAbstraction.RenderObject(::Dict{Symbol,Any}, ::GLMakie.GLVisualize.GLVisualizeShader, ::GLMakie.GLAbstraction.StandardPrerender, ::Nothing, ::Observables.Observable{GeometryTypes.HyperRectangle{3,Float32}}, ::Nothing) at /Users/anshul/.julia/dev/GLMakie/src/GLAbstraction/GLTypes.jl:336
 [5] assemble_robj(::Dict{Symbol,Any}, ::GLMakie.GLVisualize.GLVisualizeShader, ::Observables.Observable{GeometryTypes.HyperRectangle{3,Float32}}, ::UInt32, ::Nothing, ::Nothing) at /Users/anshul/.julia/dev/GLMakie/src/GLVisualize/utils.jl:38
 [6] assemble_shader(::Dict{Symbol,Any}) at /Users/anshul/.julia/dev/GLMakie/src/GLVisualize/utils.jl:65
 [7] (::GLMakie.var"##84#85"{Heatmap{...}})(::Dict{Symbol,Any}) at /Users/anshul/.julia/dev/GLMakie/src/drawing_primitives.jl:256
 [8] (::GLMakie.var"##52#55"{GLMakie.var"##84#85"{Heatmap{...}},GLMakie.Screen,Scene,Heatmap{...}})() at /Users/anshul/.julia/dev/GLMakie/src/drawing_primitives.jl:43
 [9] get!(::GLMakie.var"##52#55"{GLMakie.var"##84#85"{Heatmap{...}},GLMakie.Screen,Scene,Heatmap{...}}, ::Dict{UInt64,GLMakie.GLAbstraction.RenderObject}, ::UInt64) at ./dict.jl:452
 [10] cached_robj!(::GLMakie.var"##84#85"{Heatmap{...}}, ::GLMakie.Screen, ::Scene, ::Heatmap{...}) at /Users/anshul/.julia/dev/GLMakie/src/drawing_primitives.jl:36
 [11] draw_atomic at /Users/anshul/.julia/dev/GLMakie/src/drawing_primitives.jl:249 [inlined]
 [12] insert!(::GLMakie.Screen, ::Scene, ::Heatmap{...}) at /Users/anshul/.julia/dev/GLMakie/src/drawing_primitives.jl:110
 [13] insertplots!(::GLMakie.Screen, ::Scene) at /Users/anshul/.julia/dev/GLMakie/src/screen.jl:59
 [14] backend_display(::GLMakie.Screen, ::Scene) at /Users/anshul/.julia/dev/GLMakie/src/screen.jl:96
 [15] backend_display at /Users/anshul/.julia/dev/GLMakie/src/gl_backend.jl:58 [inlined]
 [16] display(::AbstractPlotting.PlotDisplay, ::Scene) at /Users/anshul/.julia/dev/AbstractPlotting/src/display.jl:45
```

**Comment by SimonDanisch** [comment]
Sorry, needs: https://github.com/JuliaPlots/GLMakie.jl/pull/48

File: src/conversions.jl [hunk 1]
```diff
@@ -134,6 +134,25 @@ function convert_arguments(::PointBased, positions::NTuple{N, AbstractVector}) w
     PlotSpec(points, tickranges = xyrange, ticklabels = labels)
 end
 
+function convert_arguments(
+        SL::SurfaceLike,
+        x::AbstractVector, y::AbstractVector, z::AbstractMatrix{<: Number}
```
Comment (asinghvi17) [inline src/conversions.jl]:
```suggestion
        x::AbstractVector, y::AbstractVector, z::AbstractMatrix

File: src/conversions.jl [hunk 2]
```diff
@@ -214,7 +233,7 @@ outputs them in a Tuple.
 
 `P` is the plot Type (it is optional).
 """
-function convert_arguments(::SurfaceLike, x::AbstractVecOrMat, y::AbstractVecOrMat, z::AbstractMatrix)
+function convert_arguments(::SurfaceLike, x::AbstractVecOrMat{<: Number}, y::AbstractVecOrMat{<: Number}, z::AbstractMatrix{<: Number})
```
Comment (asinghvi17) [inline src/conversions.jl]:
```suggestion
function convert_arguments(::SurfaceLike, x::AbstractVecOrMat{<: Number}, y::AbstractVecOrMat{<: Number}, z::AbstractMatrix)

File: src/conversions.jl (hunk 1 above)
Comment (SimonDanisch) [reply src/conversions.jl]: Did you test the suggested change?

Comment (asinghvi17) [reply src/conversions.jl]: Will test again, I don't remember to be honest

**Comment by asinghvi17** [comment]
```julia
image(x, y, rand(RGBAf0, 3, 3))
ERROR: StackOverflowError:
Stacktrace:
 [1] splitprec(::Type{Float64}, ::Int128) at ./twiceprecision.jl:23
 [2] TwicePrecision at ./twiceprecision.jl:197 [inlined]
 [3] steprangelen_hp(::Type{Float64}, ::Tuple{Int128,Int128}, ::Tuple{Int128,Int128}, ::Int64, ::Int64, ::Int64) at ./twiceprecision.jl:320
 [4] _linspace(::Type{Float64}, ::Int64, ::Int64, ::Int64, ::Int64) at ./twiceprecision.jl:653
 [5] _linspace at ./twiceprecision.jl:642 [inlined]
 [6] _range at ./range.jl:398 [inlined]
 [7] #range#39 at ./range.jl:88 [inlined]
 [8] #range at ./none:0 [inlined]
 [9] to_linspace at /Users/anshul/.julia/dev/AbstractPlotting/src/conversions.jl:186 [inlined]
 [10] convert_arguments(::AbstractPlotting.SurfaceLike, ::IntervalSets.Interval{:closed,:closed,Int64}, ::IntervalSets.Interval{:closed,:closed,Int64}, ::Array{ColorTypes.RGBA{Float32},2}) at /Users/anshul/.julia/dev/AbstractPlotting/src/conversions.jl:195
 [11] convert_arguments(::AbstractPlotting.SurfaceLike, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}, ::Array{ColorTypes.RGBA{Float32},2}) at /Users/anshul/.julia/dev/AbstractPlotting/src/conversions.jl:145
 ... (the last 2 lines are repeated 19979 more times)
 [39970] #convert_arguments#221(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(convert_arguments), ::Type{Image{...}}, ::Array{String,1}, ::Vararg{Any,N} where N) at /Users/anshul/.julia/dev/AbstractPlotting/src/conversions.jl:53
 [39971] convert_arguments(::Type{Image{...}}, ::Array{String,1}, ::Array{String,1}, ::Array{ColorTypes.RGBA{Float32},2}) at /Users/anshul/.julia/dev/AbstractPlotting/src/conversions.jl:51
 [39972] #plot!#208(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(plot!), ::Scene, ::Type{Image{...}}, ::Attributes, ::Array{String,1}, ::Vararg{Any,N} where N) at /Users/anshul/.julia/dev/AbstractPlotting/src/interfaces.jl:542
 [39973] plot!(::Scene, ::Type{Image{...}}, ::Attributes, ::Array{String,1}, ::Array{String,1}, ::Array{ColorTypes.RGBA{Float32},2}) at /Users/anshul/.julia/dev/AbstractPlotting/src/interfaces.jl:529
 [39974] #image#93(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(image), ::Array{String,1}, ::Vararg{Any,N} where N) at /Users/anshul/.julia/dev/AbstractPlotting/src/recipes.jl:15
 [39975] image(::Array{String,1}, ::Array{String,1}, ::Vararg{Any,N} where N) at /Users/anshul/.julia/dev/AbstractPlotting/src/recipes.jl:13
```

**Pull request merged**